c/tskit/trees.c

```diff
@@ -2387,20 +2433,20 @@ get_mutation_samples(
     if (ret != 0) {
         goto out;
     }
-
+    get_all_samples_bits(&all_samples_bits, num_samples);
     ret = tsk_tree_init(&tree, ts, TSK_NO_SAMPLE_COUNTS);
     if (ret != 0) {
         goto out;
     }
-
-    // A future improvement could get a union of all sample sets
-    // instead of all samples
-    get_all_samples_bits(&all_samples_bits, num_samples);
+    tree.index = -1; // TODO: forcing seek_from_null
 
     // Traverse down each tree, recording all samples below each mutation. We perform one
-    // preorder traversal per mutation.
-    mut_offset = 0;
-    for (ret = tsk_tree_first(&tree); ret == TSK_TREE_OK; ret = tsk_tree_next(&tree)) {
+    // preorder traversal per mutation. We start at the tree containing the first site of
+    // interest. We rely on our sites bounds checking to access the first site position
+    site_offset = 0;
+    site_idx = 0;
+    for (ret = tsk_tree_seek(&tree, ts->tables->sites.position[sites[0]], 0);
```

Comment:
Is this correct? Won't you seek to the first tree over and over doing this? See below for suggested restructuring of flow control.

Reply:
You're right about that. I didn't realize that this would cause the loop to wrap around until I messed with it in isolation. See my detailed comment below.